cli/command/telemetry_docker.go

```diff
@@ -0,0 +1,121 @@
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+const otelConfigFieldName = "otel"
+
+// dockerExporterOTLPEndpoint retrieves the OTLP endpoint used for the docker reporter
+// from the current context.
+func dockerExporterOTLPEndpoint(cli Cli) (endpoint string, secure bool, err error) {
+	meta, err := cli.ContextStore().GetMetadata(cli.CurrentContext())
+	if err != nil {
+		return "", false, err
+	}
+
+	var otelCfg any
+	switch m := meta.Metadata.(type) {
+	case DockerContext:
+		otelCfg = m.AdditionalFields[otelConfigFieldName]
+	case map[string]any:
+		otelCfg = m[otelConfigFieldName]
+	}
+
+	if otelCfg == nil {
+		return "", false, nil
+	}
+
+	otelMap, ok := otelCfg.(map[string]any)
+	if !ok {
+		return "", false, errors.Errorf(
+			"unexpected type for field %q: %T (expected: %T)",
+			otelConfigFieldName,
+			otelCfg,
+			otelMap,
+		)
+	}
+
+	// keys from https://opentelemetry.io/docs/concepts/sdk-configuration/otlp-exporter-configuration/
+	endpoint, ok = otelMap["OTEL_EXPORTER_OTLP_ENDPOINT"].(string)
+	if !ok {
+		return "", false, nil
+	}
+
+	// Parse the endpoint. The docker config expects the endpoint to be
+	// in the form of a URL to match the environment variable, but this
+	// option doesn't correspond directly to WithEndpoint.
+	//
+	// We pretend we're the same as the environment reader.
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", false, errors.Errorf("docker otel endpoint is invalid: %s", err)
+	}
+
+	switch u.Scheme {
+	case "unix":
+		// Unix sockets are a bit weird. OTEL seems to imply they
+		// can be used as an environment variable and are handled properly,
+		// but they don't seem to be as the behavior of the environment variable
+		// is to strip the scheme from the endpoint, but the underlying implementation
+		// needs the scheme to use the correct resolver.
+		//
+		// We'll just handle this in a special way and add the unix:// back to the endpoint.
+		endpoint = fmt.Sprintf("unix://%s", path.Join(u.Host, u.Path))
+	case "https":
+		secure = true
+		fallthrough
+	case "http":
+		endpoint = path.Join(u.Host, u.Path)
+	}
+	return endpoint, secure, nil
+}
+
+func dockerSpanExporter(ctx context.Context, cli Cli) (sdktrace.SpanExporter, error) {
+	endpoint, secure, err := dockerExporterOTLPEndpoint(cli)
+	if endpoint == "" || err != nil {
+		return nil, err
+	}
+
+	opts := []otlptracegrpc.Option{
+		otlptracegrpc.WithEndpoint(endpoint),
+	}
+	if !secure {
+		opts = append(opts, otlptracegrpc.WithInsecure())
+	}
+
+	exp, err := otlptracegrpc.New(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return exp, nil
+}
+
+func dockerMetricExporter(ctx context.Context, cli Cli) (sdkmetric.Exporter, error) {
+	endpoint, secure, err := dockerExporterOTLPEndpoint(cli)
+	if endpoint == "" || err != nil {
+		return nil, err
+	}
```

Comment:
I'm not sure about this. Wouldn't we want to first check the error, then return it and on another if check the endpoint as non-empty then return a custom error?
```suggestion
	endpoint, secure, err := dockerExporterOTLPEndpoint(cli)
	if err != nil {
		return nil, err
	}
	if endpoint == "" {
	        return nil, errors.New("endpoint is empty")
	}
```

Reply:
If the endpoint is empty, the intention is to return `nil, nil` since it's not an error but it's also not configured. The resulting code ends up being:

```go
if err != nil {
    return nil, err
} else if endpoint == "" {
    return nil, nil
}
```

Which is effectively the same thing as `endpoint == "" || err != nil`. I can add a comment to make that more clear.

Reply:
@Benehiko pls note that otel has it's own defaults that it uses when the value is not set (see [here](https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/))

Reply:
Oh i see! Thank you :)
I just wanted to make sure since it's not common to return `nil, nil` in Go.